RecoTracker/LSTCore/interface/alpaka/Constants.h

```diff
@@ -0,0 +1,183 @@
+  typedef float FPX;
+#endif
+
+  using Idx = alpaka_common::Idx;
+  using Dim = alpaka_common::Dim3D;
+  using Dim1d = alpaka_common::Dim1D;
+  using Vec = alpaka_common::Vec3D;
+  using Vec1d = alpaka_common::Vec1D;
+  using WorkDiv = alpaka_common::WorkDiv3D;
+
+  using Acc = ALPAKA_ACCELERATOR_NAMESPACE::Acc3D;
+  using Dev = ALPAKA_ACCELERATOR_NAMESPACE::Device;
+  using DevHost = ALPAKA_ACCELERATOR_NAMESPACE::DevHost;
+  using QueueAcc = ALPAKA_ACCELERATOR_NAMESPACE::Queue;
+
+  Vec const elementsPerThread(Vec::all(static_cast<Idx>(1)));
+
+// Needed for files that are compiled by g++ to not throw an error.
+// uint4 is defined only for CUDA, so we will have to revisit this soon when running on other backends.
+#if !defined(ALPAKA_ACC_GPU_CUDA_ENABLED) && !defined(ALPAKA_ACC_GPU_HIP_ENABLED)
+  struct uint4 {
+    unsigned int x;
+    unsigned int y;
+    unsigned int z;
+    unsigned int w;
+  };
+#endif
+
+  // Buffer type for allocations where auto type can't be used.
+  template <typename TDev, typename TData>
+  using Buf = alpaka::Buf<TDev, TData, Dim1d, Idx>;
+
+  // Allocation wrapper function to make integration of the caching allocator easier and reduce code boilerplate.
+  template <typename T, typename TAcc, typename TSize, typename TQueue>
+  ALPAKA_FN_HOST ALPAKA_FN_INLINE Buf<alpaka::Dev<TAcc>, T> allocBufWrapper(TAcc const& devAccIn,
+                                                                            TSize nElements,
+                                                                            TQueue queue) {
+#ifdef CACHE_ALLOC
+    return cms::alpakatools::allocCachedBuf<T, Idx>(devAccIn, queue, Vec1d(static_cast<Idx>(nElements)));
+#else
+    return alpaka::allocBuf<T, Idx>(devAccIn, Vec1d(static_cast<Idx>(nElements)));
+#endif
+  }
+
+  // Second allocation wrapper function when queue is not given. Reduces code boilerplate.
+  template <typename T, typename TAcc, typename TSize>
+  ALPAKA_FN_HOST ALPAKA_FN_INLINE Buf<alpaka::Dev<TAcc>, T> allocBufWrapper(TAcc const& devAccIn, TSize nElements) {
+    return alpaka::allocBuf<T, Idx>(devAccIn, Vec1d(static_cast<Idx>(nElements)));
+  }
+
+  // Wrapper function to reduce code boilerplate for defining grid/block sizes.
+  ALPAKA_FN_HOST ALPAKA_FN_INLINE Vec createVec(int x, int y, int z) {
+    return Vec(static_cast<Idx>(x), static_cast<Idx>(y), static_cast<Idx>(z));
+  }
+
+  // Adjust grid and block sizes based on backend configuration
+  template <typename Vec>
+  ALPAKA_FN_HOST ALPAKA_FN_INLINE WorkDiv createWorkDiv(const Vec& blocksPerGrid,
+                                                        const Vec& threadsPerBlock,
+                                                        const Vec& elementsPerThreadArg) {
+    Vec adjustedBlocks = blocksPerGrid;
+    Vec adjustedThreads = threadsPerBlock;
+
+    // Serial execution, so all launch parameters set to 1.
+#if defined(ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED)
+    adjustedBlocks = Vec::all(static_cast<Idx>(1));
```

Comment:
Here and below, you don't need to set the grid size to 1, the alpaka execution will run the blocks one after the other automatically.

Of course you _can_ set the grid size to 1 - the difference will be in how the kernel are executed over the elements.

Either way, I think whatever approach you choose should be documented.

Reply:
> Either way, I think whatever approach you choose should be documented.

Thinking a bit more, I'd include the reason to not use `cms::alpakatools::make_workdiv()` to this documentation.

Reply:
Uhm, I think that there may have been historically a misunderstanding of what `ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED` does.
Is this actually serial without any parallelization or does it parallelize (like `ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLED`)? ... do we have a TBB variant backend support?

The following comments are in the context of this being actually parallelized.

The long term idea was to figure out the work division on CPU separately from what was tuned on GPU.
What's the current practice for alpaka in CMS; is there a way to adjust work division based on more specific backend type? (down to hardware detail, like total SMs per card)

Here specifically the starting point was to have a single-thread execution as a default for CPU.
Initially to have a better understanding and control over the timing measurements (the standalone scripts parallelize by running multiple events in flight).

What's a good way to limit concurrency for intra-event parallelism in alpaka CPU/host backends?

In mkfit we use 
```C++
  if (limitConcurrency_) {
    tbb::task_arena arena(1);
    arena.execute(std::move(lambda));
  } else {
    tbb::this_task_arena::isolate(std::move(lambda));
  }
```

Reply:
> Uhm, I think that there may have been historically a misunderstanding of what `ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED` does. Is this actually serial without any parallelization or does it parallelize (like `ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLED`)? 

The "serial backend", i.e. `ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED`, is sequential over the execution of grid-blocks and block-threads. The thread-element dimension may get vectorized (this is up to the compiler).

> ... do we have a TBB variant backend support?

Presently, no. There are some seeds for TBB backend support in `#ifdef`s here and there, but e.g. build rules are not implemented (there are also some deeper challenges in enabling a second backend for some platform which I won't dive in here).

(Presently it is also unclear if TBB backend would bring us any benefits beyond easier debugging of concurrent execution than on GPUs. For example the standalone pixel tracking demonstrates quite significant performance degradation already between 1-thread serial processes and 1-thread TBB processes, and the throughput gets noticeably worse when increasing the number of threads per process in a setup where one process processes one event at a time and the number of processes times the number threads/process is kept constant. I _believe_ the cause has not been investigated much.)

> The following comments are in the context of this being actually parallelized.
> 
> The long term idea was to figure out the work division on CPU separately from what was tuned on GPU. What's the current practice for alpaka in CMS; is there a way to adjust work division based on more specific backend type? (down to hardware detail, like total SMs per card)

The [`cms::alpakatools::make_workdiv()`](https://github.com/cms-sw/cmssw/blob/29ab9c802e7d765e343be3cdb163e135043f93c8/HeterogeneousCore/AlpakaInterface/interface/workdivision.h#L45-L77) uses its second argument ("number of elements per block") for the number of elements on Serial backend (setting the number of blocks to 1), and for the number threads on CUDA and HIP backends (setting the number of elements to 1). Beyond that there is no established practice. I recall @fwyzard has had some ideas about it.

(I'm personally concerned if very detailed level of tuning is really worth of the effort and complexity. For HLT I could believe it _might_ be,  in which case at least the final tuning should be done in the context of the full application.)

> Here specifically the starting point was to have a single-thread execution as a default for CPU. Initially to have a better understanding and control over the timing measurements (the standalone scripts parallelize by running multiple events in flight).
> 
> What's a good way to limit concurrency for intra-event parallelism in alpaka CPU/host backends?

I'd think using the work division would be the easiest and most general way. If we would support the TBB backend, the `task_arena` could be used to limit concurrency for it, but that wouldn't work for any other CPU parallel backend (if we'd ever enable one).

Reply:
> The long term idea was to figure out the work division on CPU separately from what was tuned on GPU.
> What's the current practice for alpaka in CMS; is there a way to adjust work division based on more specific backend type? (down to hardware detail, like total SMs per card)

There's a work in progress by a PhD student, expected to be complete by the end of the year, to introduce a simple way to configure the kernel launch parameters for different backends.
The plan is to support different configurations for things like
  - GPUs vs CPUs;
  - NVIDIA GPUs vs AMD GPUs;
  - NVIDIA GPUs with `sm_86` vs `sm_75`;
  - GPU model names.

Please do let us know if you think we should add other parameters to the matching.

Reply:
> (I'm personally concerned if very detailed level of tuning is really worth of the effort and complexity. For HLT I could believe it _might_ be, in which case at least the final tuning should be done in the context of the full application.)

Part of that same work is an autotuner to find good working parameters for a given hardware configuration.

Reply:
> > What's a good way to limit concurrency for intra-event parallelism in alpaka CPU/host backends?
> 
> I'd think using the work division would be the easiest and most general way. If we would support the TBB backend, the `task_arena` could be used to limit concurrency for it, but that wouldn't work for any other CPU parallel backend (if we'd ever enable one).

I think I'm confused here 🤔 

As long as we only have the serial back-end, there is no intra-module parallelism (intra-event can still happen as we can run different modules in parallel).

If we introduce a TBB-parallel backend, a good way to limit the concurrency will be to use a small number of blocks when launching the kernels.